File: tests/test_attitude_log.py
import csv

from serial_link import AttitudeSample
from apps.attitude_log import AttitudeCsvLogger


def test_logger_writes_header_and_rows(tmp_path):
    path = tmp_path / "attitude.csv"
    samples = [
        AttitudeSample(seq=0, t_ms=0, roll=1.0, pitch=-2.0, yaw=3.0),
        AttitudeSample(seq=1, t_ms=100, roll=12.5, pitch=-3.25, yaw=180.0),
    ]
    with AttitudeCsvLogger(str(path)) as log:
        for i, s in enumerate(samples):
            log.write(s, rx_unix=1000.0 + i)
        assert log.count == 2

    with open(path, newline="") as f:
        rows = list(csv.DictReader(f))

    assert [r["seq"] for r in rows] == ["0", "1"]
    assert rows[1]["t_ms"] == "100"
    assert float(rows[1]["roll"]) == 12.5
    assert float(rows[1]["yaw"]) == 180.0
    assert rows[0]["rx_unix"] == "1000.000000"


def test_rows_flushed_before_close(tmp_path):
    # Each write flushes, so data survives even without a clean close.
    path = tmp_path / "partial.csv"
    log = AttitudeCsvLogger(str(path))
    log.write(AttitudeSample(seq=5, t_ms=500, roll=0, pitch=0, yaw=0))
    # do NOT close — read what's on disk
    with open(path, newline="") as f:
        rows = list(csv.DictReader(f))
    assert len(rows) == 1 and rows[0]["seq"] == "5"
    log.close()

File: tests/test_attitude_log.cpp
// Mirrors tests/test_attitude_log.py — CSV write then read back, optional
// columns, and the required-column / empty-file errors.
#include <gtest/gtest.h>

#include <cstdio>
#include <fstream>
#include <string>

#include "serial_link/attitude_log.hpp"

using namespace serial_link;

namespace {
std::string temp_path(const char* name) {
    return std::string(std::getenv("TMPDIR") ? std::getenv("TMPDIR") : "/tmp") + "/slcpp_" +
           name + ".csv";
}
void write_file(const std::string& path, const std::string& body) {
    std::ofstream(path) << body;
}
}  // namespace

TEST(AttitudeLog, WriteThenReadRoundtrip) {
    std::string path = temp_path("rt");
    {
        AttitudeCsvLogger log(path);
        log.write(codec::AttitudeSample{0, 0, 1.0f, 2.0f, 3.0f}, 1000.0);
        log.write(codec::AttitudeSample{1, 100, -4.5f, 5.5f, 6.5f}, 1000.1);
        EXPECT_EQ(log.count(), 2u);
    }
    auto rows = read_attitude_csv(path);
    ASSERT_EQ(rows.size(), 2u);
    EXPECT_EQ(rows[0].seq.value(), 0u);
    EXPECT_EQ(rows[1].t_ms.value(), 100u);
    EXPECT_NEAR(rows[1].roll, -4.5f, 1e-4);
    EXPECT_NEAR(rows[1].yaw, 6.5f, 1e-4);
    std::remove(path.c_str());
}

TEST(AttitudeLog, OptionalColumnsAbsent) {
    std::string path = temp_path("optional");
    write_file(path, "roll,pitch,yaw\n1.0,2.0,3.0\n");
    auto rows = read_attitude_csv(path);
    ASSERT_EQ(rows.size(), 1u);
    EXPECT_FALSE(rows[0].seq.has_value());
    EXPECT_FALSE(rows[0].t_ms.has_value());
    EXPECT_NEAR(rows[0].pitch, 2.0f, 1e-4);
    std::remove(path.c_str());
}

TEST(AttitudeLog, MissingRequiredColumnThrows) {
    std::string path = temp_path("missing");
    write_file(path, "roll,pitch\n1.0,2.0\n");
    EXPECT_THROW(read_attitude_csv(path), std::runtime_error);
    std::remove(path.c_str());
}

TEST(AttitudeLog, EmptyFileThrows) {
    std::string path = temp_path("empty");
    write_file(path, "roll,pitch,yaw\n");
    EXPECT_THROW(read_attitude_csv(path), std::runtime_error);
    std::remove(path.c_str());
}
